Sources/NIOExtras/LineBasedFrameDecoder.swift

```diff
@@ -0,0 +1,86 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2018 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+///
+/// A decoder that splits incoming `ByteBuffer`s around line end
+/// character(s) (`'\n'` or `'\r\n'`).
+///
+/// Let's, for example, consider the following received buffer:
+///
+///     +----+-------+------------+
+///     | AB | C\nDE | F\r\nGHI\n |
+///     +----+-------+------------+
+///
+/// A instance of `LineBasedFrameDecoder` will split this buffer
+/// as follows:
+///
+///     +-----+-----+-----+
+///     | ABC | DEF | GHI |
+///     +-----+-----+-----+
+///
+public class LineBasedFrameDecoder: ByteToMessageDecoder {
+    
+    public typealias InboundIn = ByteBuffer
+    public typealias InboundOut = ByteBuffer
+    public var cumulationBuffer: ByteBuffer?
+    // we keep track of the last scan end index if we didn't find the delimiter
+    private var lastIndex: ByteBufferView.Index? = nil
+    
+    public init() { }
+    
+    public func decode(ctx: ChannelHandlerContext, buffer: inout ByteBuffer) throws -> DecodingState {
+        if let frame = try findNextFrame(buffer: &buffer) {
+            ctx.fireChannelRead(wrapInboundOut(frame))
+            return .continue
+        } else {
+            return .needMoreData
+        }
+    }
+    
+    private func findNextFrame(buffer: inout ByteBuffer) throws -> ByteBuffer? {
+        var view = buffer.readableBytesView
+        // get the view's true start, end indexes
+        let _startIndex = view.startIndex
+        let _endIndex = view.endIndex
+        // start where we left off last scan or from the beginning
+        let firstIndex = min(self.lastIndex ?? _startIndex, _endIndex)
```

Comment:
It is not safe to assume the indices of a `ByteBuffer` are static across invocations of `decode`. Instead of saving the explicit index, you should save the offset of that index from the `readerIndex`. That offset will stay valid.

Reply:
In fact, it's probably a good idea to write a test that does this. It's easily done: write a test that has one and a half lines in the byte buffer, such that it returns the first and then partially completes the second. Then, call `discardReadBytes` on the cumulation buffer and save it back to the handler, before you write the rest of the data in.